pkg/controller/atlasproject/ipaccess_list.go

```diff
@@ -59,194 +44,166 @@ func ensureIPAccessList(ctx context.Context, workflowCtx *workflow.Context, akoP
+func syncIPAccessList(ctx context.Context, service *workflow.Context, projectID string, current, desired []project.IPAccessList) error {
+	currentMap := map[string]project.IPAccessList{}
+	for _, item := range current {
+		currentMap[genIPAccessListKey(item)] = item
 	}
 
-	if _, _, err := client.ProjectIPAccessList.Create(context.Background(), projectID, operatorAccessLists); err != nil {
-		return workflow.Terminate(workflow.ProjectIPNotCreatedInAtlas, err.Error())
+	desiredMap := map[string]project.IPAccessList{}
+	for _, item := range desired {
+		desiredMap[genIPAccessListKey(item)] = item
 	}
-	return workflow.OK()
-}
 
-func deleteIPAccessFromAtlas(client mongodbatlas.Client, projectID string, listsToRemove []set.Identifiable, log *zap.SugaredLogger) error {
-	for _, l := range listsToRemove {
-		if _, err := client.ProjectIPAccessList.Delete(context.Background(), projectID, l.Identifier().(string)); err != nil {
+	for key, ipAccessList := range currentMap {
+		if _, ok := desiredMap[key]; ok {
+			continue
+		}
+
+		_, err := service.Client.ProjectIPAccessList.Delete(ctx, projectID, mapToEntryValue(ipAccessList, false))
+		if err != nil {
 			return err
 		}
-		log.Debugw("Removed IPAccessList from Atlas as it's not specified in current AtlasProject", "id", l.Identifier())
 	}
-	return nil
-}
 
-type IPAccessListStatusType string
+	toCreate := make([]*mongodbatlas.ProjectIPAccessList, 0, len(desired))
+	for key, ipAccessList := range desiredMap {
+		if _, ok := currentMap[key]; !ok {
+			toCreate = append(
+				toCreate,
+				&mongodbatlas.ProjectIPAccessList{
+					AwsSecurityGroup: ipAccessList.AwsSecurityGroup,
+					CIDRBlock:        ipAccessList.CIDRBlock,
+					Comment:          ipAccessList.Comment,
+					DeleteAfterDate:  ipAccessList.DeleteAfterDate,
+					GroupID:          projectID,
+					IPAddress:        ipAccessList.IPAddress,
+				},
+			)
+		}
+	}
+
+	if len(toCreate) == 0 {
+		return nil
+	}
 
-const (
-	IPAccessListActive  IPAccessListStatusType = "ACTIVE"
-	IPAccessListFailed  IPAccessListStatusType = "FAILED"
-	IPAccessListPending IPAccessListStatusType = "PENDING"
-)
+	_, _, err := service.Client.ProjectIPAccessList.Create(ctx, projectID, toCreate)
 
-type IPAccessListStatus struct {
-	Status string `json:"STATUS"`
+	return err
 }
 
-// getAccessListEntry returns the identifier for the accessList. It should be exactly one of IPAddress, CIDRBlock
-// or AwsSecurityGroup. This function assumes that the accessList is already validated and has only one of these
-// fields populated.
-func getAccessListEntry(accessList mongodbatlas.ProjectIPAccessList) string {
-	if accessList.IPAddress != "" {
-		return url.QueryEscape(accessList.IPAddress)
+func mapToEntryValue(ipAccessList project.IPAccessList, escape bool) string {
+	entry := ""
+
+	switch {
+	case ipAccessList.CIDRBlock != "":
+		entry = ipAccessList.CIDRBlock
+		quads := strings.Split(ipAccessList.CIDRBlock, "/")
+		if quads[1] == "32" {
+			entry = quads[0]
```

Comment:
Do we assume here that CIDRBlock is correct? Otherwise `quads` requires length check

Reply:
Yes, I assumed Atlas will not make it wrong (or themselves would fail) and added validating on our side [see this](https://github.com/mongodb/mongodb-atlas-kubernetes/pull/1121/files#diff-16a1c718d0fd0dbe812a2687880f7406b57bf39e6cbd24620d600a48adc9930cR212)